### PR TITLE
Apply range-filters in get when probing active-listener cache

### PR DESCRIPTION
**Pull request by IanWyszynski**

### CODE REVIEW

File: packages/database/src/core/SyncTree.ts [hunk 1]
```diff
@@ -505,6 +506,42 @@ export class SyncTree {
...
+      let serverCacheNode: CacheNode | null = new CacheNode(
+        serverCache,
+        true,
+        false
+      );
+      let writesCache: WriteTreeRef | null = this.pendingWriteTree_.childWrites(
+        query.path
+      );
+      let view: View = syncPoint.getView(query, writesCache, serverCache, true);
+      return view.getCompleteServerCache(query.path);
```
Comment (IanWyszynski) [inline packages/database/src/core/SyncTree.ts:540]: @schmidt-sebastian This seems to apply the filters for startAfter/endBefore, but it erroneously returns null for this test: https://github.com/firebase/firebase-js-sdk/blob/master/packages/database/test/query.test.ts#L3159-L3179.

Can you have a look at this? I'm trying to figure out why this is happening but not making much progress.

File: packages/database/src/core/SyncPoint.ts [hunk 2]
```diff
@@ -94,23 +94,12 @@ export class SyncPoint {
...
+  getView(
     query: Query,
-    eventRegistration: EventRegistration,
     writesCache: WriteTreeRef,
     serverCache: Node | null,
     serverCacheComplete: boolean
-  ): Event[] {
+  ): View {
     const queryId = query.queryIdentifier();
     let view = this.views.get(queryId);
```
Comment (schmidt-sebastian) [inline packages/database/src/core/SyncPoint.ts:104]: As some lint tooling says: 'view' is never reassigned. Use 'const' instead.

Or assign it in the if block.

File: packages/database/src/core/SyncPoint.ts [hunk 3]
```diff
@@ -140,10 +129,37 @@ export class SyncPoint {
...
+   * @return {!Array.<!Event>} Events to raise.
+   */
+  addEventRegistration(
+    query: Query,
+    eventRegistration: EventRegistration,
+    writesCache: WriteTreeRef,
+    serverCache: Node | null,
+    serverCacheComplete: boolean
+  ): Event[] {
+    let view = this.getView(
```
Comment (schmidt-sebastian) [inline packages/database/src/core/SyncPoint.ts:154]: s/let/const

File: packages/database/src/core/SyncTree.ts [hunk 4]
```diff
@@ -505,6 +506,47 @@ export class SyncTree {
...
+      syncPoint = new SyncPoint();
+      this.syncPointTree_ = this.syncPointTree_.set(path, syncPoint);
+    } else {
+      foundAncestorDefaultView =
+        foundAncestorDefaultView || syncPoint.hasCompleteView();
+      serverCache = serverCache || syncPoint.getCompleteServerCache(Path.Empty);
+    }
+    if (serverCache != null) {
+      let serverCacheNode: CacheNode | null = new CacheNode(
+        serverCache,
```
Comment (schmidt-sebastian) [inline packages/database/src/core/SyncTree.ts:532]: Please fix these lint failures.

File: packages/database/test/query.test.ts [hunk 5]
```diff
@@ -1399,6 +1399,38 @@ describe('Query Tests', () => {
     expect(Object.values(snap.val())).to.deep.equal([snap.val()[childOne.key]]);
   });
 
+  it('Ensure startAfter on key index works with overlapping listener', async () => {
+    const node = getRandomNode() as Reference;
+    const childOne = node.push();
+    const childTwo = node.push();
+    await childOne.set(1);
+    await childTwo.set(2);
```
Comment (schmidt-sebastian) [inline packages/database/test/query.test.ts:1407]: Do the tests still pass if you don't `await` here? I would like to confirm that we correctly apply local writes.

Comment (IanWyszynski) [reply packages/database/test/query.test.ts]: They do still pass.

File: packages/database/src/core/SyncTree.ts [hunk 6]
```diff
@@ -505,6 +506,47 @@ export class SyncTree {
     );
   }
 
+  getServerValue(query: Query): Node | null {
+    const path = query.path;
+    let serverCache: Node | null = null;
+    let foundAncestorDefaultView = false;
```
Comment (schmidt-sebastian) [inline packages/database/src/core/SyncTree.ts:512]: `foundAncestorDefaultView` is not used and can be removed.

**Review by schmidt-sebastian — COMMENTED** [review]
This needs a changelog entry in response to https://firebase.google.com/support/release-notes/js#realtime-database_5

File: packages/database/test/query.test.ts [hunk 7]
```diff
@@ -1399,6 +1399,38 @@ describe('Query Tests', () => {
     expect(Object.values(snap.val())).to.deep.equal([snap.val()[childOne.key]]);
   });
 
+  it('Ensure startAfter on key index works with overlapping listener', async () => {
+    const node = getRandomNode() as Reference;
+    const childOne = node.push();
+    const childTwo = node.push();
+    await childOne.set(1);
+    childTwo.set(2);
```
Comment (schmidt-sebastian) [inline packages/database/test/query.test.ts:1407]:
```suggestion
    // Create a server synced and a latency-compensated write
    await childOne.set(1);
    childTwo.set(2);
```

File: packages/database/test/query.test.ts [hunk 8]
```diff
@@ -1399,6 +1399,38 @@ describe('Query Tests', () => {
...
+    expect(Object.keys(snap.val())).to.deep.equal([childTwo.key]);
+    expect(Object.values(snap.val())).to.deep.equal([snap.val()[childTwo.key]]);
+  });
+
+  it('Ensure endBefore on key index works with overlapping listener', async () => {
+    const node = getRandomNode() as Reference;
+    const childOne = node.push();
+    const childTwo = node.push();
+    await childOne.set(1);
+    childTwo.set(2);
```
Comment (schmidt-sebastian) [inline packages/database/test/query.test.ts:1423]:
```suggestion
    // Create a server synced and a latency-compensated write
    await childOne.set(1);
    childTwo.set(2);
```

**Comment by IanWyszynski** [comment]
@Feiyang1 for code-owner review.

File: .changeset/small-icons-allow.md [hunk 9]
```diff
@@ -0,0 +1,4 @@
+---
+'@firebase/database': patch
+---
+Fixed an issue with Query.get() where Query filters are not applied to data in some cases.
```
Comment (schmidt-sebastian) [inline .changeset/small-icons-allow.md:4]:
```suggestion
Fixed an issue with `Query.get()` where Query filters are not applied to data in some cases.
```

File: packages/database/src/core/SyncPoint.ts [hunk 10]
```diff
@@ -128,10 +130,37 @@ export class SyncPoint {
...
 
+  /**
+   * Add an event callback for the specified query.
+   *
+   * @param {!Query} query
+   * @param {!EventRegistration} eventRegistration
+   * @param {!WriteTreeRef} writesCache
+   * @param {?Node} serverCache Complete server cache, if we have it.
+   * @param {boolean} serverCacheComplete
+   * @return {!Array.<!Event>} Events to raise.
```
Comment (schmidt-sebastian) [inline packages/database/src/core/SyncPoint.ts:146]:
```suggestion
   * @param query
   * @param eventRegistration
   * @param writesCache
   * @param serverCache Complete server cache, if we have it.
   * @param serverCacheComplete
   * @return Events to raise.
```

I just removed the Closure types in the rest of the SDK.

File: packages/database/src/core/view/View.ts [hunk 11]
```diff
@@ -95,6 +95,14 @@ export class View {
...
   }
 
+  getCompleteNode(): Node | null {
+    return this.viewCache_.getCompleteEventSnap();
+  }
+
+  /**
+   * @param {!Path} path
+   * @return {?Node}
+   */
```
Comment (schmidt-sebastian) [inline packages/database/src/core/view/View.ts:105]:
```suggestion
```

**Review by schmidt-sebastian — APPROVED** [review]
Additional tests LGTM. Thanks for all the extra effort!

**Pull request merged**